finagle-http/src/main/scala/com/twitter/finagle/http/HeaderMap.scala

```diff
@@ -52,67 +54,88 @@ abstract class HeaderMap
   def += (kv: (String, Date)): HeaderMap =
     += ((kv._1, HeaderMap.format(kv._2)))
 
-  override def empty: HeaderMap = new MapHeaderMap(mutable.Map.empty)
+  override def empty: HeaderMap = MapHeaderMap()
 }
 
+private[finagle] case class HeaderValuePair(header: String, value: String) {
+  val canonicalName = HeaderValuePair.canonicalName(header)
+}
+
+private[finagle] object HeaderValuePair {
+  def canonicalName(header: String) = header.toLowerCase
```

Comment:
the default encoding will be the platform if unspecified, so we should use `c.t.io.Charsets.UsAscii` i believe.